### DISCUSSION TITLE
Allows to register OPs in the CPU backend.

### DISCUSSION THREAD

**Discussion by Djip007** [ggml-org/llama.cpp#10350] [Ideas]
Create a full backend is a lot of [work](https://github.com/ggerganov/llama.cpp/pull/10343#issuecomment-2480834278), Some actual backend only need to implement some optimised OP, like BLAS, AMX, ... or [tinyBLAS](https://github.com/ggerganov/llama.cpp/issues/10183) 

I have some test for use the RDNA3 iGPU of AMD CPU (AMD Ryzen 9 7940HS w/ Radeon 780M Graphics)
I want to create some hight speed gemm for FP8 on CPU, but for good speed we need to use more "classique" matmul kernel, like the 5 level BLIS structure.
When XDNA driver will be available on linux I'll like to have a look on this NPU.

A full backend is nice for discrete accelerator, but to much work (copy) for integreted accelerator that use CPU memory.

Next there is some "idea" to build multiple CPU backend and use the "best" supported for the current CPU.

So my feeling is that is may be good to have the possibility to "register" OP, and select at runtime those which are possible, and the best for the current compute. May be let the user chose some of them...

**Comment by slaren** [comment]
The way we handle this is with `ggml_backend_sched`. That's how the BLAS and AMX backends work despite only implementing matrix multiplication.

**Reply by Djip007** [reply]
I know that, I use it for tinyBLAS and you rejected it  https://github.com/ggerganov/llama.cpp/pull/10343#issuecomment-2480834278

look @ggerganov is agrees with that: https://github.com/ggerganov/llama.cpp/issues/10183#issuecomment-2480961965

I understand why not to make a backend for these cases. 
I don't know if you already have an idea to reintegrate the AMX backend into the CPU Backend.

**Reply by slaren** [reply]
The plan to reintegrate the AMX backend would be to create a new buffer type that converts the weights to the layout that the AMX backend needs them, and then check in the matrix multiplication the buffer type to determine if the AMX matrix multiplication code should be used. Basically extending the same that is done in #9921 for the aarch64 types.

**Reply by ggerganov** [reply]
@slaren I created this issue https://github.com/ggerganov/llama.cpp/issues/10359 to track this. Feel free to update it if needed.

**Reply by Djip007** [reply]
I saw it quickly, I have to look more closely.
Is the idea not to record OPs but devices on CPU backend that support part of the OPs and optimized for "accelerators"?

**Reply by slaren** [reply]
The idea is that backends can support multiple buffer types, which represent different ways of storing tensors. In this case, the CPU backend would support the standard CPU buffer type that stores tensors in the standard ggml layout, but it can also work with repacked weights that are stored in the aarch64 buffer type, or the future AMX buffer type. The backend is responsible for checking the buffer type of the tensors and use that to determine which code needs to be used. Llama.cpp can automatically use the best buffer type available to load the weights.
I don't think we need to add more devices to the CPU backend, the single device is enough. Accelerators such as NPUs or iGPUs should be implemented as a separate backend, however, support for instruction sets should be kept in the CPU backend, in the same way we handle it for AVX, NEON, etc.

**Reply by Djip007** [reply]
How do you thing to generalise 
https://github.com/ggerganov/llama.cpp/blob/ce2e59ba107cf71ed566040ff20a15d1c58e09c2/ggml/src/ggml-cpu/ggml-cpu.cpp#L459-L469

with AMX matmul_op we will have 1 more buffer type.
and it may needed to add some more for most optimised OP (conv, fp8_gemm, fp8_gemv ... )

 > Llama.cpp can automatically use the best buffer type available to load the weights.
It can become very complicated to decide where to put a buffer, it will depend on the OP which must or can be used. 

https://github.com/ggerganov/llama.cpp/blob/ce2e59ba107cf71ed566040ff20a15d1c58e09c2/ggml/src/ggml-cpu/ggml-cpu.cpp#L75-L79
https://github.com/ggerganov/llama.cpp/blob/ce2e59ba107cf71ed566040ff20a15d1c58e09c2/ggml/src/ggml-cpu/ggml-cpu-aarch64.c#L3545-L3559

How do you know it need to be repack, it is only for now for matmul OP, what if the same tensor type is use for other OP?

**Reply by slaren** [reply]
llama.cpp calls `supports_op` with the op that will be used with the weight during loading to determine what buffer type to use. This was added in #10026, there are more details there. If the `supports_op` function of the CPU backend becomes too complicated as more buffer types are added it can be refactored in many ways, I don't think it's a problem at the moment.

**Reply by Djip007** [reply]
I missed this pull-request/change. Thanks, I'll have a look a it.

In case, are there any plans to add all Dynamic types in ggml_type or are there plans to move it somewhere else?

**Reply by slaren** [reply]
> In case, are there any plans to add all Dynamic types in ggml_type or are there plans to move it somewhere else?

I don't know if it would be worth adding new ggml_types for the AMX repacked tensors layouts. I don't have a strong opinion, but at the moment I think it would be simpler to avoid adding new types until there is a clear reason to do that.

**Reply by Djip007** [reply]
OK I think it is to much for me to start with [AMX](https://github.com/ggerganov/llama.cpp/issues/10359)
I will try to add a simple bf16 matmul with repacking

I totally agree with not adding any new type.
It is use here:

https://github.com/ggerganov/llama.cpp/blob/ce2e59ba107cf71ed566040ff20a15d1c58e09c2/ggml/src/ggml-cpu/ggml-cpu.c#L410-L417
Have you any plan with that?

same "question" with that: isn't it better to move this to the cpu backend ?
https://github.com/ggerganov/llama.cpp/blob/ce2e59ba107cf71ed566040ff20a15d1c58e09c2/ggml/src/ggml-common.h#L206-L210

What do you think if I move the gemv/gemm on another array and use some sort of "CPU_TENSOR_TYPE"? 

Note: for me this gemv/gemm look like a way to register some "OP"

**Reply by slaren** [reply]
I am not sure, maybe it would be good to phase out the Q4_0_x_x file types, remove them from `llama-quantize` and eventually make them available only via online repacking of Q4_0 models. There are some advantages to having a Q4_0_x_x model file that does not require conversion, it allows using mmap and reduces the load time, but I am not sure if that is significant enough to keep them. If you want to do some refactoring of the way the CPU backend handles this with a new struct or some other way go for it, anything that improves the code quality is welcome.

**Reply by Djip007** [reply]
OK I'll try something and see if it work.

**Reply by Djip007** [reply]
Create a draft https://github.com/ggerganov/llama.cpp/pull/10446 with my current testing...
More need to be done :sweat_smile:
